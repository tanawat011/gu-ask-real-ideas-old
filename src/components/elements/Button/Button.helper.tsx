import clsx from 'clsx'

import { BUTTON_STYLE } from '@constants'
import { PropButton } from '.'

export const renderPrimaryClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-primary-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-primary-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-primary-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-primary-default text-primary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-primary-default text-primary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-primary-default-flat text-primary-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-primary-default', classHover())
  }

  return clsx(
    'bg-primary-default text-white',
    classHover(),
    !isLoading && 'active:bg-primary-pressed',
    hasShadow && 'shadow-nextui shadow-primary-default',
  )
}

export const renderSecondaryClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-secondary-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-secondary-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-secondary-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-secondary-default text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-secondary-default text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx(
      'bg-secondary-default-flat text-secondary-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-secondary-default', classHover())
  }

  return clsx(
    'bg-secondary-default text-white',
    classHover(),
    !isLoading && 'active:bg-secondary-pressed',
    hasShadow && 'shadow-nextui shadow-secondary-default',
  )
}

export const renderSuccessClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-success-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-success-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-success-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-success-default text-success-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-success-default text-success-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-success-default-flat text-success-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-success-default', classHover())
  }

  return clsx(
    'bg-success-default text-white',
    classHover(),
    !isLoading && 'active:bg-success-pressed',
    hasShadow && 'shadow-nextui shadow-success-default',
  )
}

export const renderWarningClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-warning-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-warning-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-warning-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-warning-default text-warning-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-warning-default text-warning-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-warning-default-flat text-warning-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-warning-default', classHover())
  }

  return clsx(
    'bg-warning-default text-white',
    classHover(),
    !isLoading && 'active:bg-warning-pressed',
    hasShadow && 'shadow-nextui shadow-warning-default',
  )
}

export const renderInfoClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-info-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-info-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-info-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-info-default text-info-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-info-default text-info-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-info-default-flat text-info-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-info-default', classHover())
  }

  return clsx(
    'bg-info-default text-white',
    classHover(),
    !isLoading && 'active:bg-info-pressed',
    hasShadow && 'shadow-nextui shadow-info-default',
  )
}

export const renderDangerClass = (props: PropButton): string => {
  const { isLoading, isDisabled, hasShadow } = props

  if (isDisabled) {
    switch (props.buttonStyle) {
      case BUTTON_STYLE.LIGHT:
        return 'disabled:bg-transparent text-nextui-gray-600'

      default:
        return 'disabled:bg-danger-disabled text-nextui-gray-600'
    }
  }

  const classHover = () => {
    if (!isLoading) {
      switch (props.buttonStyle) {
        case BUTTON_STYLE.OUTLINE:
        case BUTTON_STYLE.FLAT:
        case BUTTON_STYLE.LIGHT:
          return 'hover:opacity-80'

        case BUTTON_STYLE.GHOST:
          return 'hover:text-white hover:bg-danger-default'

        case BUTTON_STYLE.DEFAULT:
        default:
          return 'hover:bg-danger-hover'
      }
    }
  }

  if (props.buttonStyle === BUTTON_STYLE.OUTLINE) {
    return clsx(
      'bg-transparent border-solid border-2 border-danger-default text-danger-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.GHOST) {
    return clsx(
      'bg-transparent border-solid border-2 border-danger-default text-danger-default',
      classHover(),
    )
  }

  if (props.buttonStyle === BUTTON_STYLE.FLAT) {
    return clsx('bg-danger-default-flat text-danger-default', classHover())
  }

  if (props.buttonStyle === BUTTON_STYLE.LIGHT) {
    return clsx('bg-transparent text-danger-default', classHover())
  }

  return clsx(
    'bg-danger-default text-white',
    classHover(),
    !isLoading && 'active:bg-danger-pressed',
    hasShadow && 'shadow-nextui shadow-danger-default',
  )
}
