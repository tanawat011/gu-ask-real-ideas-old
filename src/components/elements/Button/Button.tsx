import { MouseEvent } from 'react'
import * as _AllIcon from '@heroicons/react/solid'
// import { ArchiveIcon } from '@heroicons/react/solid'

import {
  BUTTON_STYLE,
  COLOR,
  LOADING_COLOR,
  LOADING_STYLE,
  SIZE,
} from '@constants'
import { Loading } from '@element/Loader'
import { clsx } from '@libs'

// eslint-disable-next-line @typescript-eslint/no-var-requires
const AllIcon = require('@heroicons/react/solid')

export type PropButton = {
  width?: number
  label?: string
  size?: SIZE
  color?: COLOR
  buttonStyle?: BUTTON_STYLE
  icon?: keyof typeof _AllIcon | ''
  iconRight?: keyof typeof _AllIcon | ''
  iconFit?: boolean
  disabled?: boolean
  loading?: boolean
  rounded?: boolean
  rippleEffect?: boolean
  wAuto?: boolean
  shadow?: boolean
  onClick?: (event: MouseEvent) => void
}

export const Button: React.FC<PropButton> = (props) => {
  const {
    width,
    label,
    size = SIZE.MD,
    color = COLOR.PRIMARY,
    iconRight,
    icon,
    buttonStyle = BUTTON_STYLE.DEFAULT,
    iconFit = false,
    disabled = false,
    loading = false,
    rounded = false,
    rippleEffect = true,
    wAuto = false,
    shadow,
    onClick,
  } = props

  const createRippleEffect = (event: MouseEvent) => {
    const button = event.currentTarget as EventTarget &
      Element & { offsetLeft: number; offsetTop: number }

    const circle = document.createElement('span')
    const diameter = Math.max(button.clientWidth, button.clientHeight)
    const radius = diameter / 2

    circle.style.width = circle.style.height = `${diameter}px`
    circle.style.left = `${event.clientX - button.offsetLeft - radius}px`
    circle.style.top = `${event.clientY - button.offsetTop - radius}px`
    circle.classList.add('ripple')

    if (buttonStyle !== BUTTON_STYLE.DEFAULT) {
      circle.classList.add(`btn__bg-${color.toLowerCase()}`)
    } else {
      circle.classList.add('bg-black')
    }

    const ripple = button.getElementsByClassName('ripple')[0]

    if (ripple) {
      ripple.remove()
    }

    button.appendChild(circle)
  }

  const isDisabled = () => disabled || loading

  const renderCursor = () => (disabled || loading) && 'cursor-not-allowed'
  const renderDisabled = () => (disabled || loading) && 'btn__disabled'
  const renderRounded = () => rounded && 'btn__rounded'
  const renderAutoWidth = () => wAuto && 'btn__auto-w'
  const renderSize = () => `btn__${size.toLowerCase()}`

  const renderColor = () => {
    const _color = color.toLowerCase()
    const _style = buttonStyle.toLowerCase()

    if (disabled) {
      if (buttonStyle === BUTTON_STYLE.LIGHT) {
        return `btn__${_color}-light-disabled`
      }

      return `btn__${_color}-disabled`
    }

    if (buttonStyle === BUTTON_STYLE.DEFAULT) {
      return clsx(
        `btn__${_color}`,
        loading && `btn__${_color}-loading`,
        shadow && `btn__${_color}-shadow`,
      )
    }

    return clsx(
      `btn__${_color}-${_style}`,
      loading && `btn__${_color}-${_style}-loading`,
    )
  }

  const renderIcon = (_icon: string, position: 'left' | 'right') => {
    const Icon = AllIcon[_icon as keyof typeof AllIcon]

    const isLeft = position === 'left'

    const classWidthIcon = label ? 'w-8' : 'w-6'
    let classPaddingIcon = ''

    if (label) {
      classPaddingIcon = isLeft ? 'pr-2' : 'pl-2'
    }

    const classIcon = clsx(
      classWidthIcon,
      classPaddingIcon,
      iconFit && 'absolute',
      iconFit && isLeft ? 'left-4' : 'right-4',
    )

    return <Icon className={classIcon} />
  }

  const renderChild = () =>
    loading ? (
      <Loading loaderStyle={LOADING_STYLE.DASH} color={LOADING_COLOR.WHITE} />
    ) : (
      <>
        {icon && <>{renderIcon(icon, 'left')}</>}
        {label}
        {iconRight && <>{renderIcon(iconRight, 'right')}</>}
      </>
    )

  return (
    <button
      className={clsx(
        'btn',
        renderCursor(),
        renderDisabled(),
        renderRounded(),
        renderAutoWidth(),
        renderSize(),
        renderColor(),
      )}
      style={{ ...(width && { width }) }}
      disabled={isDisabled()}
      onClick={(event) => {
        if (rippleEffect) {
          createRippleEffect(event)
        }

        if (onClick) {
          onClick(event)
        }
      }}
    >
      {renderChild()}
    </button>
  )
}
