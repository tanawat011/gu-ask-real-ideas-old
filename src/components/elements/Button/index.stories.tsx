import { ComponentMeta, ComponentStory } from '@storybook/react'

import { Button } from '.'

export default {
  title: 'Button/Default',
  component: Button,
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Default = Template.bind({})

Default.args = {
  title: 'Button',
  size: 'medium',
  color: 'primary-default',
}
