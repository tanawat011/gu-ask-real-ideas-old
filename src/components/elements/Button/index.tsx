import clsx from 'clsx'
import { VFC } from 'react'
import Image from 'next/image'

interface ButtonType {
  title: string
  size: 'small' | 'medium' | 'large'
  color: 'yellow' | 'red' | 'white' | 'bg'
  icon?: string
  isDisabled?: boolean
  width?: number
}

export const Button: VFC<ButtonType> = ({
  width,
  title,
  size,
  color,
  icon,
  isDisabled = false,
}) => {
  const renderSize = (size: string) => {
    if (size === 'small') {
      return 'h-6-5 w-52-7 text-xss rounded'
    }

    if (size === 'small' && color === 'white') {
      return 'h-6 w-28-1 text-xss rounded'
    }

    if (size === 'medium') {
      return 'h-7-5 w-64-2 text-xss rounded-md'
    }

    if (size === 'medium' && color === 'white') {
      return 'h-6 w-64-8 text-xss rounded-md'
    }

    if (size === 'large') {
      return 'h-10 w-64-2 text-xs rounded-lg'
    }

    if (size === 'large' && color === 'white') {
      return 'h-6 w-64-8 text-xs rounded-lg'
    }
  }

  const renderColor = (color: string) => {
    if (color === 'yellow') {
      return clsx(
        'bg-action-default hover:bg-action-hover active:bg-action-pressed',
        isDisabled && 'disabled:bg-action-default opacity-40',
      )
    }

    if (color === 'red') {
      return clsx(
        'bg-secondary-candy hover:bg-candy-hover active:bg-candy-pressed',
        isDisabled && 'disabled:bg-secondary-candy opacity-40',
      )
    }

    if (color === 'white') {
      return clsx(
        'bg-neutral-fade-100 hover:bg-neutral-gray-200 active:bg-neutral-dark-400',
        isDisabled && 'disabled:bg-neutral-fade-100 opacity-40',
      )
    }

    if (color === 'bg') {
      return clsx(
        'bg-neutral-gray-300 hover:bg-neutral-gray-200 active:bg-neutral-dark-400',
        isDisabled && 'disabled:bg-neutral-gray-300 opacity-40',
      )
    }
  }

  return (
    <button
      className={`
        ${renderSize(size)}
        ${renderColor(color)}
        flex items-center justify-center font-medium text-neutral-black
      `}
      style={{ width }}
      disabled={isDisabled}
    >
      {icon && (
        <Image className='w-3 h-3 m-1 pointer-events-none' src={icon} alt='' />
      )}{' '}
      {title}
    </button>
  )
}
