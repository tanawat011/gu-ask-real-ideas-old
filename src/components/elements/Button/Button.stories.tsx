import clsx from 'clsx'
import { ComponentMeta, ComponentStory } from '@storybook/react'

import { BUTTON_STYLE, COLOR, SIZE } from '@constants'
import { Button } from '@element/Button'

const ButtonSizeAssembleComponent: React.FC<
  React.ComponentProps<typeof Button>
> = (props) => {
  const classCard = clsx('flex', 'justify-start', 'items-center', 'm-3')

  return (
    <div className=''>
      <div className={classCard}>
        <Button label='Mini' {...props} size={SIZE.XS} />
      </div>

      <div className={classCard}>
        <Button label='Small' {...props} size={SIZE.SM} />
      </div>

      <div className={classCard}>
        <Button label='Medium' {...props} size={SIZE.MD} />
      </div>

      <div className={classCard}>
        <Button label='Large' {...props} size={SIZE.LG} />
      </div>

      <div className={classCard}>
        <Button label='XLarge' {...props} size={SIZE.XL} />
      </div>

      <div className={classCard}>
        <Button label='Auto width' {...props} wAuto />
      </div>

      <div className={classCard}>
        <Button label='Custom width(300px)' {...props} width={300} />
      </div>

      <div className={classCard}>
        <Button label='Custom width(600px)' {...props} width={600} />
      </div>
    </div>
  )
}

const ButtonAssembleComponent: React.FC<
  React.ComponentProps<typeof Button> & { title?: string }
> = (props) => {
  const classCard = clsx(
    'flex',
    'flex-col',
    'justify-center',
    'items-center',
    'm-3',
  )

  return (
    <>
      <span className='text-white'>{props.title}</span>
      <div className='flex flex-wrap'>
        <div className={classCard}>
          <Button
            label='Primary'
            {...props}
            color={COLOR.PRIMARY}
            {...(props.loading && { width: 91.69 })}
            wAuto
          />
        </div>

        <div className={classCard}>
          <Button
            label='Secondary'
            {...props}
            color={COLOR.SECONDARY}
            {...(props.loading && { width: 110.91 })}
            wAuto
          />
        </div>

        <div className={classCard}>
          <Button
            label='Success'
            {...props}
            color={COLOR.SUCCESS}
            {...(props.loading && { width: 88.75 })}
            wAuto
          />
        </div>

        <div className={classCard}>
          <Button
            label='Warning'
            {...props}
            color={COLOR.WARNING}
            {...(props.loading && { width: 96.28 })}
            wAuto
          />
        </div>

        <div className={classCard}>
          <Button
            label='Info'
            {...props}
            color={COLOR.INFO}
            {...(props.loading && { width: 62.67 })}
            wAuto
          />
        </div>

        <div className={classCard}>
          <Button
            label='Danger'
            {...props}
            color={COLOR.DANGER}
            {...(props.loading && { width: 87.02 })}
            wAuto
          />
        </div>
      </div>
    </>
  )
}

export default {
  title: 'Button',
  component: Button,
  parameters: {
    docs: {
      description: {
        component: `Button description
          * In the Docs menu will no effect ripple when click button
        `,
      },
      page: null,
    },
  },
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (props) => (
  <Button label='Default' {...props} />
)

export const DefaultButton = Template.bind({})
DefaultButton.story = {
  name: 'Default',
}

const ButtonSizeAssembleTemplate: ComponentStory<typeof Button> = (props) => (
  <ButtonSizeAssembleComponent {...props} />
)

export const Size = ButtonSizeAssembleTemplate.bind({})

const ButtonAssembleTemplate: ComponentStory<typeof Button> = (props) => (
  <ButtonAssembleComponent {...props} />
)

export const Color = ButtonAssembleTemplate.bind({})
export const Shadow = ButtonAssembleTemplate.bind({})
Shadow.args = {
  shadow: true,
}
export const Disabled = ButtonAssembleTemplate.bind({})
Disabled.args = {
  disabled: true,
}
export const Rounded = ButtonAssembleTemplate.bind({})
Rounded.args = {
  buttonStyle: BUTTON_STYLE.OUTLINE,
  rounded: true,
}
export const OutlineStyle = ButtonAssembleTemplate.bind({})
OutlineStyle.args = {
  buttonStyle: BUTTON_STYLE.OUTLINE,
}
export const GhostStyle = ButtonAssembleTemplate.bind({})
GhostStyle.args = {
  buttonStyle: BUTTON_STYLE.GHOST,
  rounded: true,
}
export const FlatStyle = ButtonAssembleTemplate.bind({})
FlatStyle.args = {
  buttonStyle: BUTTON_STYLE.FLAT,
}
export const LightStyle = ButtonAssembleTemplate.bind({})
LightStyle.args = {
  buttonStyle: BUTTON_STYLE.LIGHT,
  rounded: true,
}

const AllButtonAssembleTemplate: ComponentStory<typeof Button> = (props) => (
  <>
    <ButtonAssembleComponent {...props} title='Default style' />
    <ButtonAssembleComponent
      {...props}
      buttonStyle={BUTTON_STYLE.OUTLINE}
      title='Outline style'
    />
    <ButtonAssembleComponent
      {...props}
      buttonStyle={BUTTON_STYLE.GHOST}
      title='Ghost style'
    />
    <ButtonAssembleComponent
      {...props}
      buttonStyle={BUTTON_STYLE.FLAT}
      title='Flat style'
    />
    <ButtonAssembleComponent
      {...props}
      buttonStyle={BUTTON_STYLE.LIGHT}
      title='Light style'
    />
  </>
)

export const AllOnlyIcons = AllButtonAssembleTemplate.bind({})
AllOnlyIcons.args = {
  label: '',
  icon: 'ArchiveIcon',
}

export const AllIcons = AllButtonAssembleTemplate.bind({})
AllIcons.args = {
  icon: 'ArchiveIcon',
}

export const AllLoading = AllButtonAssembleTemplate.bind({})
AllLoading.args = {
  loading: true,
}

export const AllDisabled = AllButtonAssembleTemplate.bind({})
AllDisabled.args = {
  disabled: true,
}

export const AllDisabledWithOnlyIcon = AllButtonAssembleTemplate.bind({})
AllDisabledWithOnlyIcon.args = {
  label: '',
  icon: 'ArchiveIcon',
  disabled: true,
}

export const AllDisabledWithIcon = AllButtonAssembleTemplate.bind({})
AllDisabledWithIcon.args = {
  icon: 'ArchiveIcon',
  disabled: true,
}
