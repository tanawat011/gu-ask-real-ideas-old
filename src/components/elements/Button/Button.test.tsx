import { render } from '@testing-library/react'

import { BUTTON_STYLE, COLOR, SIZE } from '@constants'
import { Button } from '@element/Button'

jest.mock('@heroicons/react/solid', (): unknown => ({
  ArchiveIcon: (props: never) => <svg {...props} />,
}))

describe('<Button />', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should render to match snapshot', () => {
    expect.assertions(1)

    const { asFragment } = render(
      <Button
        color={COLOR.PRIMARY}
        size={SIZE.MD}
        icon='ArchiveIcon'
        label='Button A'
        width={100}
        shadow
      />,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render without props', () => {
    expect.assertions(1)

    const { asFragment } = render(<Button label='Button A' />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render with onClick', async () => {
    expect.assertions(1)

    const onClick = jest.fn()

    const { findByRole } = render(<Button label='Button A' onClick={onClick} />)

    const button = await findByRole('button')

    button.click()

    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('should be rounded button', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button rounded />)

    const button = await findByRole('button')

    expect(button).toHaveClass(`btn__rounded`)
  })

  it(`should be shadow button`, async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button shadow />)

    const button = await findByRole('button')

    expect(button).toHaveClass(`btn__primary-shadow`)
  })

  it(`should be shadow button, loading button`, async () => {
    expect.assertions(3)

    const { findByRole } = render(<Button shadow loading />)

    const button = await findByRole('button')

    expect(button).toHaveClass(`btn__primary-shadow`)
    expect(button).toHaveClass('cursor-not-allowed')
    expect(button.children[0].children[0].nodeName).toBe('svg')
  })

  it('should be auto width button', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button wAuto />)

    const button = await findByRole('button')

    expect(button).toHaveClass(`btn__auto-w`)
  })

  it('should be icon button', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button label='Test' icon='ArchiveIcon' />)

    const button = await findByRole('button')

    expect(button.children[0].nodeName).toBe('svg')
  })

  it('should be icon button without label', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button icon='ArchiveIcon' />)

    const button = await findByRole('button')

    expect(button.children[0].nodeName).toBe('svg')
  })

  it('should be icon button, fit icon button', async () => {
    expect.assertions(1)

    const { findByRole } = render(
      <Button label='Test' icon='ArchiveIcon' iconFit />,
    )

    const button = await findByRole('button')

    expect(button.children[0]).toHaveClass('absolute')
  })

  it('should be right icon button', async () => {
    expect.assertions(1)

    const { findByRole } = render(
      <Button label='Test' iconRight='ArchiveIcon' />,
    )

    const button = await findByRole('button')

    expect(button.children[0].nodeName).toBe('svg')
  })

  it('should be right icon button without label', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button iconRight='ArchiveIcon' />)

    const button = await findByRole('button')

    expect(button.children[0].nodeName).toBe('svg')
  })

  it('should be right icon button, fit icon button', async () => {
    expect.assertions(1)

    const { findByRole } = render(
      <Button label='Test' iconRight='ArchiveIcon' iconFit />,
    )

    const button = await findByRole('button')

    expect(button.children[0]).toHaveClass('absolute')
  })

  it('should be icon button, right icon button', async () => {
    expect.assertions(2)

    const { findByRole } = render(
      <Button label='Test' icon='ArchiveIcon' iconRight='ArchiveIcon' />,
    )

    const button = await findByRole('button')

    expect(button.children[0].nodeName).toBe('svg')
    expect(button.children[1].nodeName).toBe('svg')
  })

  it('should be icon button, right icon button, fit icon button', async () => {
    expect.assertions(2)

    const { findByRole } = render(
      <Button
        label='Test'
        icon='ArchiveIcon'
        iconRight='ArchiveIcon'
        iconFit
      />,
    )

    const button = await findByRole('button')

    expect(button.children[0]).toHaveClass('absolute')
    expect(button.children[1]).toHaveClass('absolute')
  })

  it('should be icon button, right icon button, fit icon button without label', async () => {
    expect.assertions(2)

    const { findByRole } = render(
      <Button icon='ArchiveIcon' iconRight='ArchiveIcon' iconFit />,
    )

    const button = await findByRole('button')

    expect(button.children[0]).toHaveClass('absolute')
    expect(button.children[1]).toHaveClass('absolute')
  })

  it('should be fit icon button', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button label='Test' iconFit />)

    const button = await findByRole('button')

    expect(button).toBeTruthy()
  })

  it('should be fit icon button without label', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button iconFit />)

    const button = await findByRole('button')

    expect(button).toBeTruthy()
  })

  it('should be without ripple effect button', async () => {
    expect.assertions(1)

    const { asFragment, findByRole } = render(
      <Button label='Test' rippleEffect={false} />,
    )

    const button = await findByRole('button')

    button.click()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should be custom width button', async () => {
    expect.assertions(1)

    const { findByRole } = render(<Button label='Test' width={300} />)

    const button = await findByRole('button')

    expect(button).toHaveStyle('width: 300px')
  })

  Object.keys(SIZE).forEach((size) => {
    it(`should be ${size.toLowerCase()} button`, async () => {
      expect.assertions(1)

      const { findByRole } = render(
        <Button size={SIZE[size as keyof typeof SIZE]} />,
      )

      const button = await findByRole('button')

      expect(button).toHaveClass(`btn__${size.toLowerCase()}`)
    })
  })

  Object.keys(COLOR).forEach((color) => {
    Object.keys(BUTTON_STYLE).forEach((style) => {
      const _postfixStyle = style === 'DEFAULT' ? '' : `-${style.toLowerCase()}`
      const disabledClass =
        style === 'LIGHT'
          ? `btn__${color.toLowerCase()}-light-disabled`
          : `btn__${color.toLowerCase()}-disabled`

      it(`should be ${color.toLowerCase()} button, ${style.toLowerCase()} button`, async () => {
        expect.assertions(1)

        const { findByRole } = render(
          <Button
            color={COLOR[color as keyof typeof COLOR]}
            buttonStyle={BUTTON_STYLE[style as keyof typeof BUTTON_STYLE]}
          />,
        )

        const button = await findByRole('button')

        expect(button).toHaveClass(
          `btn__${color.toLowerCase()}${_postfixStyle}`,
        )
      })

      it(`should be ${color.toLowerCase()} button, ${style.toLowerCase()} button, disabled button`, async () => {
        expect.assertions(2)

        const { findByRole } = render(
          <Button
            color={COLOR[color as keyof typeof COLOR]}
            buttonStyle={BUTTON_STYLE[style as keyof typeof BUTTON_STYLE]}
            disabled
          />,
        )

        const button = await findByRole('button')

        expect(button).toHaveClass('btn__disabled')
        expect(button).toHaveClass(disabledClass)
      })

      it(`should be ${color.toLowerCase()} button, ${style.toLowerCase()} button, loading button`, async () => {
        expect.assertions(2)

        const { findByRole } = render(
          <Button
            color={COLOR[color as keyof typeof COLOR]}
            buttonStyle={BUTTON_STYLE[style as keyof typeof BUTTON_STYLE]}
            loading
          />,
        )

        const button = await findByRole('button')

        expect(button.children[0].children[0].nodeName).toBe('svg')
        expect(button).toHaveClass(
          `btn__${color.toLowerCase()}${_postfixStyle}-loading`,
        )
      })

      it(`should be ${color.toLowerCase()} button, ${style.toLowerCase()} button, ripple effect button`, async () => {
        expect.assertions(2)

        const { findByRole } = render(
          <Button
            color={COLOR[color as keyof typeof COLOR]}
            buttonStyle={BUTTON_STYLE.OUTLINE}
            label='Test'
            rippleEffect
          />,
        )

        const button = await findByRole('button')

        button.click()
        button.click() // Click again for test old ripple effect removed and re-assign

        expect(button.children[0]).toHaveClass('ripple')
        expect(button.children[0]).toHaveClass(`btn__bg-${color.toLowerCase()}`)
      })
    })
  })
})
