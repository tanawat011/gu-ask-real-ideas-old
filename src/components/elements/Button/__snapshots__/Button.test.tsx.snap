// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should be without ripple effect button 1`] = `
<DocumentFragment>
  <button
    class="btn btn__md btn__primary"
  >
    Test
  </button>
</DocumentFragment>
`;

exports[`<Button /> should render to match snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn btn__md btn__primary btn__primary-shadow"
    style="width: 100px;"
  >
    <svg
      class="w-8 pr-2 right-4"
    />
    Button A
  </button>
</DocumentFragment>
`;

exports[`<Button /> should render without props 1`] = `
<DocumentFragment>
  <button
    class="btn btn__md btn__primary"
  >
    Button A
  </button>
</DocumentFragment>
`;
